janus_server/src/hpke.rs

```diff
@@ -0,0 +1,379 @@
+
+        let config = HpkeConfig {
+            id: HpkeConfigId(0),
+            kem_id: HpkeKemId::X25519HkdfSha256,
+            kdf_id: HpkeKdfId::HkdfSha512,
+            aead_id: HpkeAeadId::ChaCha20Poly1305,
+            public_key: HpkePublicKey(public_key.to_bytes().as_slice().to_vec()),
+        };
+
+        Self {
+            task_id,
+            config,
+            label,
+            sender_role,
+            recipient_role,
+            recipient_private_key: HpkePrivateKey(private_key.to_bytes().as_slice().to_vec()),
+        }
+    }
+
+    /// Decrypt `ciphertext` and return the plaintext.
+    ///
+    /// In PPM, an HPKE context can only be used once (we have no means of
+    /// ensuring that sender and recipient "increment" nonces in lockstep), so
+    /// this method consumes self.
+    fn open(self, ciphertext: &HpkeCiphertext, associated_data: &[u8]) -> Result<Vec<u8>, Error> {
+        // We must manually dispatch to each possible specialization of open
+        match (self.config.kem_id, self.config.kdf_id, self.config.aead_id) {
+            (HpkeKemId::P256HkdfSha256, HpkeKdfId::HkdfSha256, HpkeAeadId::Aes256Gcm) => {
+                open::<AesGcm256, HkdfSha256, DhP256HkdfSha256>(
+                    HpkeApplicationInfo::new(
+                        self.task_id,
+                        self.label,
+                        self.sender_role,
+                        self.recipient_role,
+                    ),
+                    ciphertext,
+                    associated_data,
+                    &self.recipient_private_key,
+                )
+            }
+            (HpkeKemId::X25519HkdfSha256, HpkeKdfId::HkdfSha512, HpkeAeadId::ChaCha20Poly1305) => {
+                open::<ChaCha20Poly1305, HkdfSha512, X25519HkdfSha256>(
+                    HpkeApplicationInfo::new(
+                        self.task_id,
+                        self.label,
+                        self.sender_role,
+                        self.recipient_role,
+                    ),
+                    ciphertext,
+                    associated_data,
+                    &self.recipient_private_key,
+                )
+            }
+            (_, _, _) => Err(Error::InvalidConfiguration(
+                "unsupported set of HPKE algorithms",
+            )),
+        }
+    }
+}
+
+// This function exists separately from struct HpkeRecipient to abstract away its
+// generic parameters
+fn open<Encrypt: Aead, Derive: Kdf, Encapsulate: Kem>(
+    application_info: HpkeApplicationInfo,
+    ciphertext: &HpkeCiphertext,
+    associated_data: &[u8],
+    serialized_recipient_private_key: &HpkePrivateKey,
+) -> Result<Vec<u8>, Error> {
+    // Deserialize recipient priv into the appropriate PrivateKey type for
+    // the KEM
+    let recipient_private_key =
+        Encapsulate::PrivateKey::from_bytes(&serialized_recipient_private_key.0)?;
+
+    // Deserialize sender encapsulated pub into the appropriate EncappedKey
```

Comment:
nit: this would be an encapsulated context/encapsulated key/encapsulated KEM secret.
```suggestion
    // Deserialize sender encapsulated key into the appropriate EncappedKey
```